### PR TITLE
Add cotire

### PR DESCRIPTION
**Pull request by jakirkham**
Adds a recipe for [`cotire`]( https://github.com/sakra/cotire ) (written from scratch). This is useful for getting a unity build up and running quickly. May be useful to have this around for some of our longer builds.

cc @jschueller @frol @ivoflipse @Korijn @patricksnape

### CODE REVIEW

**Comment by jakirkham** [comment]
Feedback and maintainers welcome.

**Comment by msarahan** [comment]
Looks useful.  I probably don't need to be a maintainer, but I'll back you up if you want.  The recipe looks fine to me - as long as the CIs come back OK.

File: recipes/cotire/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    #
+    # https://github.com/sakra/cotire/pull/100
+    #
+    - PR_100.diff
+
+build:
+  number: 0
+  script:
+    - cp "${SRC_DIR}/CMake/cotire.cmake" "${PREFIX}"/share/cmake*/Modules               # [unix]
+    - copy "%SRC_DIR%\\CMake\\cotire.cmake" "%LIBRARY_PREFIX%"\\share\\cmake*\\Modules  # [win]
```
Comment (jakirkham) [inline recipes/cotire/meta.yaml]: Am I doing something wrong here, @msarahan? It seem to have issues copying this on Windows. I really don't know the batch syntax.

File: recipes/cotire/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    - cmake
+
+  run:
+    - cmake
+
+test:
+  commands:
+    # Verify that the cotire CMake file is in place.
+    - test -f "${PREFIX}"/share/cmake*/Modules/cotire.cmake                           # [unix]
+    - if not exist "%LIBRARY_PREFIX%"\\share\\cmake*\\Modules\\cotire.cmake exit 1    # [win]
```
Comment (jakirkham) [inline recipes/cotire/meta.yaml]: Also, want to verify that this works on Windows too. I'm hoping it does, but I don't know.

**Comment by jakirkham** [comment]
No worries. I'm guessing others might get excited about this.

For the most part, maintaining this is mind-numbingly simple so I'm hoping its just not a big deal.

File: recipes/cotire/meta.yaml (hunk 1 above)
Comment (jakirkham) [inline recipes/cotire/meta.yaml]: Maybe the quotes are causing this issues. Will try dropping them.

File: recipes/cotire/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+test:
+  requires:
+     - python 2.7.*  # [win and py27]
+     - python 3.4.*  # [win and py34]
+     - python 3.5.*  # [win and py35]
+  commands:
+    # Verify that the cotire CMake file is in place.
+    - test -f "${PREFIX}"/share/cmake*/Modules/cotire.cmake                           # [unix]
+    - if not exist %LIBRARY_PREFIX%\\share\\cmake*\\Modules\\cotire.cmake exit 1      # [win]
```
Comment (jakirkham) [inline recipes/cotire/meta.yaml]: Ugh...Windows doesn't like the wildcard here. Any suggestions?

Comment (jakirkham) [inline recipes/cotire/meta.yaml]: NVM guess I can `cd` here too.

File: recipes/cotire/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    - cd %LIBRARY_PREFIX%\\share\\cmake*\\Modules && if not exist cotire.cmake exit 1  # [win]
+
+
+    # Use cotire in a simple test to make sure it works.
+    - cd "${SRC_DIR}/src" && mkdir build && cd build                                  # [unix]
+    - cd "%SRC_DIR%/src" && mkdir build && cd build                                   # [win]
+    - cmake -G "Unix Makefiles" ..                                                    # [unix]
+    - cmake -G "NMake Makefiles" ..                                                   # [win]
+    # Cannot find `cl` on Windows during tests.
+    - cmake --build .                                                                 # [unix]
```
Comment (jakirkham) [inline recipes/cotire/meta.yaml]: Have to skip on Windows. See this issue ( https://github.com/conda/conda-build/issues/1059 ).

File: recipes/cotire/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+source:
+  fn: cotire-{{ version }}.tar.gz
+  url: https://github.com/sakra/cotire/archive/cotire-{{ version }}.tar.gz
+  sha256: {{ sha256 }}
+  patches:
+    # Include cotire in example CMakeLists.txt.
+    #
+    # https://github.com/sakra/cotire/pull/100
+    #
+    - PR_100.diff
```
Comment (jakirkham) [inline recipes/cotire/meta.yaml]: Patch from this PR ( https://github.com/sakra/cotire/pull/100 ). Needed to build the example.

File: recipes/cotire/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    - cd "${SRC_DIR}/src" && mkdir build && cd build                                  # [unix]
+    - cd "%SRC_DIR%/src" && mkdir build && cd build                                   # [win]
+    - cmake -G "Unix Makefiles" ..                                                    # [unix]
+    # Cannot find `cl` on Windows during tests.
+    # Please see this issue.
+    #
+    # https://github.com/conda/conda-build/issues/1059
+    #
+    #- cmake -G "NMake Makefiles" ..                                                  # [win]
+    - cmake --build .                                                                 # [unix]
```
Comment (jakirkham) [inline recipes/cotire/meta.yaml]: Had to skip building on Windows. Would be a nice test to include on all platforms, but it seems a bit challenging to get Windows to work. See this issue ( https://github.com/conda/conda-build/issues/1059 ).

**Comment by jakirkham** [comment]
This is now passing. 😄

**Comment by Korijn** [comment]
Awesome. Shame about the Windows testing bug. Anyway, LGTM!

**Comment by jschueller** [comment]
Nice, I think this kind of things tends to be included in projects though rather than used as an optional client library. At least that's what we did for openturns. Or maybe your approcah is to patch the target package yourself to use cotire ?

**Comment by jakirkham** [comment]
Yeah, I figured it might make it easier for people to play with. Also, this way it can be versioned separately, which should reduce some maintenance burden on devs (or that was my thinking at least).

**Pull request merged**